src/go/rpk/pkg/cli/cmd/root.go

```diff
@@ -165,6 +201,324 @@ func tryExecPlugin(h pluginHandler, args []string) (string, error) {
 	return foundPath, h.exec(foundPath, args[len(pieces):])
 }
 
+const pluginPrefix = "rpk-"                // does not support --help-autocomplete
+const pluginPrefixAutoComplete = "rpk.ac-" // supports --help-autocomplete
+
+// A plugin is made up of the "pieces" that are used to call it, and the binary
+// path for the plugin on disk. Reversing the documentation on tryExecPlugin,
+// if a plugin filepath is /foo/bar/rpk.ac-baz_boz-biz, then the pieces will be
+// baz-boz biz.
+type plugin struct {
+	pieces []string
+	path   string
+}
+
+// listPlugins returns all plugins found in fs across the given search
+// directories.
+//
+// Unlike kubectl, we do not allow plugins to be tacked on paths within other
+// plugins. That is, we do not allow rpk-foo-bar to be an additional plugin on
+// top of rpk-foo.
+//
+// We do support plugins defining themselves as "rpk-foo-bar", even though that
+// reserves the "foo" plugin namespace.
+func listPlugins(fs afero.Fs, searchDirs []string) []plugin {
+	searchDirs = uniqueTrimmedStrs(searchDirs)
+
+	uniquePlugins := make(map[string]int) // plugin name (e.g., mm3 or cloud) => index into plugins
+	var plugins []plugin
+	for _, searchDir := range searchDirs {
+		infos, err := afero.ReadDir(fs, searchDir)
+		if err != nil {
+			log.Debugf("unable to read dir %s from PATH: %v", searchDir, err)
+			continue
+		}
+		for _, info := range infos {
+			if info.IsDir() {
+				continue
+			}
+
+			name := info.Name()
+			if !strings.HasPrefix(name, pluginPrefix) && !strings.HasPrefix(name, pluginPrefixAutoComplete) {
+				continue
+			}
+
+			fullPath := filepath.Join(searchDir, name)
+
+			if info.Mode()&0111 == 0 {
+				log.Debugf("matching plugin path %s is not executable, skipping", fullPath)
+				continue
+			}
+
+			// Reverse our args-to-plugin-binary logic to get the
+			// "pieces" that are used to call this plugin.
+			if strings.HasPrefix(name, pluginPrefix) {
+				name = strings.TrimPrefix(name, pluginPrefix)
+			} else {
+				name = strings.TrimPrefix(name, pluginPrefixAutoComplete)
+			}
+			if len(name) == 0 { // e.g., "rpk-"
+				log.Debugf("invalid empty plugin name at path %s", fullPath)
+				continue
+			}
+
+			pieces := toPieces(name)
+			pluginName := pieces[0]
+
+			if priorAt, exists := uniquePlugins[pluginName]; exists {
+				log.Debugf("skipping duplicate plugin %s at path %s, because it is previously overshadowed by path %s", pluginName, fullPath, plugins[priorAt].path)
+				continue
+			}
+
+			uniquePlugins[name] = len(plugins)
+			plugins = append(plugins, plugin{
+				pieces: pieces,
+				path:   fullPath,
+			})
+
+		}
+	}
+
+	return plugins
+}
+
+// Converts a command name to its argument pieces.
+func toPieces(command string) []string {
+	pieces := strings.Split(command, "-")
+	for i, piece := range pieces {
+		pieces[i] = strings.ReplaceAll(piece, "_", "-")
```

Comment:
I don't really have a strong opinion here, I just followed the `kubectl` syntax to start. After writing all of this a dash feels natural to me, but at this point I'm biased. Maybe worth getting a third opinion here?

Reply:
I think that your suggestion is great because it keeps the logic simpler and still supports dashes and spaces. I'm not sure why kubernetes went the two-characters-replaced route when a similar one-character-replaced route option exists.